### PR TITLE
table function view

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add table function `view` which turns an subquery into a table object. This helps passing queries around. For instance, it can be used in remote/cluster table functions.

Detailed description / Documentation draft:

Document needs update. Nothing special.

### CODE REVIEW

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
I'd like to make code more generic to avoid
`name == "view"`

File: src/Parsers/ASTFunction.cpp [hunk 1]
```diff
@@ -109,6 +109,18 @@ void ASTFunction::formatImplWithoutAlias(const FormatSettings & settings, Format
     nested_need_parens.need_parens = true;
     nested_dont_need_parens.need_parens = false;
 
+    if (name == "view")
```
Comment (alexey-milovidov) [inline src/Parsers/ASTFunction.cpp:112]: Maybe just remove this code completely? What will be worse?

File: src/Parsers/ExpressionElementParsers.cpp [hunk 2]
```diff
@@ -309,6 +309,44 @@ bool ParserFunction::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
     return true;
 }
 
+bool ParserTableFunction::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
```
Comment (alexey-milovidov) [inline src/Parsers/ExpressionElementParsers.cpp:312]: Maybe just allow to have query as an argument to any function?
And then throw exception on query analysis stage if it's wrong place...

Thus we can avoid having special case for "view" here.

File: src/Parsers/ExpressionElementParsers.cpp [hunk 3]
```diff
@@ -1319,7 +1357,7 @@ bool ParserExpressionElement::parseImpl(Pos & pos, ASTPtr & node, Expected & exp
         || ParserRightExpression().parse(pos, node, expected)
         || ParserCase().parse(pos, node, expected)
         || ParserColumnsMatcher().parse(pos, node, expected) /// before ParserFunction because it can be also parsed as a function.
-        || ParserFunction().parse(pos, node, expected)
+        || ParserTableFunction().parse(pos, node, expected)
```
Comment (alexey-milovidov) [inline src/Parsers/ExpressionElementParsers.cpp:1360]: This looks very strange...

**Review by alexey-milovidov — APPROVED** [review]
```
Alexey Milovidov, [02.09.20 15:43]
Ok. About table function view - maybe better to treat query as function argument (put to arguments) rather than special query field in ASTFunction?

Amos Bird, [02.09.20 15:44]
[In reply to Alexey Milovidov]
but that will complex all function visitors...

Alexey Milovidov, [02.09.20 15:44]
Ok, then will keep it as is.
```

**Comment by simPod** [comment]
I read tests but don't very understand what is this good for 🤔

**Comment by alexey-milovidov** [comment]
It's intended to pass arbitrary query as is to several remote servers via composition of `remote`/`cluster` and `view` functions.

**Comment by lelezhan** [comment]
So when we could use the view function? I tested it, it also executes subquery in view for local shard firstly,then executes remote shard. Compared with distribute table, can it improve the performance of query execution when combining it with clusters function, like WITH CTE subquery

**Pull request merged**